### ISSUE TITLE
Unity's Animation Rigging package: works in parts already

### ISSUE THREAD

**Issue by LarsJandel** [KybernetikGames/animancer#14]
Tried applying `Animancer` to the Ninja from the Siggraph 2019 workshop project using Unity's `Animation Rigging` package:
https://github.com/Unity-Technologies/animation-rigging-workshop-siggraph2019
---
_Edit: Error while using v3.1:_
The standard `Animancer Component` works fine, `Named` and `Hybrid` prompt the following error and do not get build in full:
```
NullReferenceException: Object reference not set to an instance of an object
Animancer.Editor.NamedAnimancerComponentEditor.DoDefaultAnimationField (UnityEditor.SerializedProperty playAutomatically) (at Assets/Plugins/Animancer/Internal/Editor Utilities/NamedAnimancerComponentEditor.cs:74)
Animancer.Editor.NamedAnimancerComponentEditor.DoOverridePropertyGUI (System.String name, UnityEditor.SerializedProperty property, UnityEngine.GUIContent label) (at Assets/Plugins/Animancer/Internal/Editor Utilities/NamedAnimancerComponentEditor.cs:31)
Animancer.Editor.AnimancerPlayableEditor.DoOtherFieldsGUI () (at Assets/Plugins/Animancer/Internal/Editor Utilities/AnimancerPlayableEditor.cs:124)
Animancer.Editor.AnimancerPlayableEditor.OnInspectorGUI () (at Assets/Plugins/Animancer/Internal/Editor Utilities/AnimancerPlayableEditor.cs:87)
UnityEditor.UIElements.InspectorElement+<>c__DisplayClass55_0.<CreateIMGUIInspectorFromEditor>b__0 () (at <83a73882c51c4602b3d34743827d03e7>:0)
UnityEngine.GUIUtility:ProcessEvent(Int32, IntPtr)
```
---
`Animation Rigging` incorporates a bunch of constraints, like `Two Bone IK` used here on the arms and legs. The character rig's animation get overridden via created control and deform rigs. In the screen-recording, only the arms have their weight set to 1, other constraints like the legs are set to 0. First I play the scene with the provided override `Controller` and an _Idle_ animation. Then I add the `Animancer` component, a _Walk_ animation and reapply the override `Controller`. This works fine. The unused `Base Layer` can be deleted - the overrides still work.
Here's a low res gif. I can send you the video clip if you want to see it in full size. Thanks.
![AnimancerAnimationRigging_200112A](https://user-images.githubusercontent.com/32457827/72217971-93f3f900-3535-11ea-90d1-8e4af1814144.gif)

**Comment by KybernetikGames** [comment]
Thanks for checking it out.

Are you using Animancer v4.0 Beta 5? Because line 74 in NamedAnimancerComponentEditor.cs is:

```
var playAutomaticallyArea = AnimancerGUI.StealFromLeft(ref area, playAutomaticallyWidth);
```

Which shouldn't be able to throw a NullReferenceException.

**Comment by LarsJandel** [comment]
...uh damn, as you are asking, I switched the project, wasn't thinking and installed straight from the asset store, so not even v4... Sorry, let me try that again. 😅

**Comment by LarsJandel** [comment]
Okay, with `v4.05b` no errors in the console. It seems like `Named` and `Hybrid` do get applied properly now, got `Named` working as expected. With `Hybrid`, I would need to figure out, how to work with that, as currently even though the override controller gets pushed to the Animancer component, the overrides from `Animation Rigging` did not work as expected anymore, but this might be a matter of setting it up differently. But as tried before, `Named` and the standard component kinda work like a hybrid in this scenario already.
Another detail if this matters: I used the scene and the Ninja as they came, using a generic rig and no avatar applied.
Now the bad part, the errors are gone but Unity did crash a couple of times. I can send you the reports if that helps.

**Comment by KybernetikGames** [comment]
I just tried it out and got a crash when setting up the character which I suspect was caused by something wrong with my setup because once I went around and made sure all the references were properly assigned it worked fine, both with named and hybrid.

Here's what I get when playing the *Run* animation with IK:

![run-with-ik](https://user-images.githubusercontent.com/43698611/73130147-89aa2280-403e-11ea-8a80-87e26e804c26.gif)

Did you manage to figure out why your hybrid setup wasn't working?

**Comment by LarsJandel** [comment]
Okay, I went back to the project, made sure to use the latest pre-release - Animancer Pro v4.0f2 - and added the Hybrid one onto the ramen eating Ninja. It works fine now 🙂. Well, I don't know, remember, what went wrong the first time.
I also created a scene from scratch with your character using the Hybrid component and Animation Rigging and that worked, too.
Well, I tried to apply the Animation Rigging Two Bone IK to your Hybrid examples, but that didn't work out of the box. But, I don't know what I am doing... For example with the _Simple Hybrid_ example, I got the following errors, but this was probably a brute force attempt that can't work and as said before, setting it up from scratch does work. ...but maybe you want to try yourself, adding the Animation Rigging to that example, or you know the cause already.
```Transform 'LowPolyMan' not found in HumanDescription.
Could not resolve 'Hips/UpperLeg.R' because the avatar is invalid. Please assign a valid Avatar or create one with AvatarBuilder.
Could not resolve 'Hips/UpperLeg.R/LowerLeg.R' because the avatar is invalid. Please assign a valid Avatar or create one with AvatarBuilder.
Could not resolve 'Hips/UpperLeg.R/LowerLeg.R/Foot.R' because the avatar is invalid. Please assign a valid Avatar or create one with AvatarBuilder.
System.InvalidOperationException: The TransformStreamHandle cannot be resolved.
Thrown from job: UnityEngine.Animations.Rigging.TwoBoneIKConstraintJob
This Exception was thrown from a job compiled with Burst, which has limited exception support. Turn off burst (Jobs -> Burst -> Enable Compilation) to inspect full exceptions & stacktraces.```

**Comment by KybernetikGames** [comment]
Try renaming the objects in the Control Rig to not have the same name as the bones in the model. I remember seeing errors or warnings like that and I'm pretty sure that's how I fixed them.

**Comment by LarsJandel** [comment]
I guess, I did sth wrong. As there is no control rig involved and there were no naming issues, I just redid the procedure and it works. Repeated two more times, all good. Sorry for this false alarm... Just one issue, you might be able to solve: Added the _Two Bone IK_ to the _Simple Hybrid_ example, I can modify the target and hint objects at runtime, but when I click any of the buttons, the IK helpers will get reset to their initial position. Is there a trick to avoid this behavior? Thanks.
Works fine = the offset sticks while playing the animations, when I modify the IK helpers before pressing play. Just wondering, if there is a trick for runtime as well.
